### PR TITLE
fix: add enroll page and handle logged in member

### PR DESCRIPTION
**Pull request by spaenleh**

### CODE REVIEW

File: cypress/e2e/redirections.cy.ts [hunk 1]
```diff
@@ -32,7 +32,7 @@ describe('Item page', () => {
       );
     });
 
-    it('Should redirect to auth with url parameter', () => {
+    it.only('Should redirect to auth with url parameter', () => {
```
Comment (ReidyT) [inline cypress/e2e/redirections.cy.ts:35]:
```suggestion
    it('Should redirect to auth with url parameter', () => {
```

File: src/langs/en.json [hunk 2]
```diff
@@ -34,5 +34,17 @@
   "MAP_BUTTON_DISABLED_TEXT": "This item does not have a geolocation",
   "FROM_SHORTCUT_BUTTON_TEXT": "Back to {{name}}",
   "GUEST_LIMITATION_TEXT": "You are currently using Graasp with the guest account <1>{{name}}</1>. In order to use all features of Graasp, you have to log out and create a Graasp account.",
-  "GUEST_SIGN_OUT_BUTTON": "Log out and Create a Graasp account"
+  "GUEST_SIGN_OUT_BUTTON": "Log out and Create a Graasp account",
+  "ENROLL_SCREEN_TITLE": "Enroll to {{name}}",
+  "ENROLL_SCREEN_DESCRIPTION": "This resource allows access with your Graasp account. To access this course, simply click Enroll",
```
Comment (ReidyT) [inline src/langs/en.json:39]: It is always a course? If not, maybe we could update it:

```suggestion
  "ENROLL_SCREEN_DESCRIPTION": "This resource allows access with your Graasp account. To access it, simply click Enroll",
```

File: src/langs/en.json [hunk 3]
```diff
@@ -34,5 +34,17 @@
...
+  "ENROLL_BUTTON_TEXT": "Enroll",
+  "REQUEST_ACCESS_PENDING_TITLE": "Your access request is pending",
+  "REQUEST_ACCESS_PENDING_DESCRIPTION": "An admin needs to approve your request for you to access this item.",
```
Comment (ReidyT) [inline src/langs/en.json:42]:
```suggestion
  "REQUEST_ACCESS_PENDING_DESCRIPTION": "An administrator must approve your request before you can access this item.",
```

File: src/modules/pages/HomePage.tsx [hunk 4]
```diff
@@ -35,17 +41,25 @@ const DisplayItems = ({
 }: {
   items?: PackedItem[];
   isLoading: boolean;
-}): JSX.Element[] | false => {
+}): JSX.Element[] | JSX.Element | false => {
   if (items) {
+    if (!items.length) {
+      return (
+        <Alert severity="info" sx={{ m: 1, width: '100%' }}>
+          You have no elements.
```
Comment (ReidyT) [inline src/modules/pages/HomePage.tsx:49]: Translate?

File: src/modules/pages/itemPage/EnrollContent.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import { Button } from '@graasp/ui';
+
+import { CircleUser } from 'lucide-react';
+
+import { usePlayerTranslation } from '@/config/i18n';
+import { mutations } from '@/config/queryClient';
+import { ENROLL_BUTTON_SELECTOR } from '@/config/selectors';
+import { PLAYER } from '@/langs/constants';
+
+// eslint-disable-next-line import/prefer-default-export
```
Comment (ReidyT) [inline src/modules/pages/itemPage/EnrollContent.tsx:12]: export default to avoid this?

File: src/modules/pages/itemPage/ItemPage.tsx [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  const { data: item, isLoading: isItemLoading } = useItem(itemId);
+  const { data: itemLoginSchemaType, isLoading: isLoadingItemLoginSchemaType } =
+    useItemLoginSchemaType({ itemId });
+  const { data: currentAccount, isLoading: isLoadingMember } =
+    useCurrentMember();
+
+  if (!itemId) {
+    navigate(HOME_PATH);
+    return null;
+  }
```
Comment (ReidyT) [inline src/modules/pages/itemPage/ItemPage.tsx:35]: Should we also consider a null item?

File: src/modules/pages/itemPage/ItemPage.tsx [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  return (
+    <ItemLoginWrapper
+      itemId={itemId}
+      item={item}
+      currentAccount={currentAccount}
+      signIn={itemLoginSignIn}
+      itemLoginSchemaType={itemLoginSchemaType}
+      usernameInputId={ITEM_LOGIN_USERNAME_INPUT_ID}
+      signInButtonId={ITEM_LOGIN_SIGN_IN_BUTTON_ID}
+      passwordInputId={ITEM_LOGIN_PASSWORD_INPUT_ID}
```
Comment (ReidyT) [inline src/modules/pages/itemPage/ItemPage.tsx:45]: Should these have to be passed as parameters or could we just use their directly in the `ItemLoginWrapper`?

File: src/modules/pages/itemPage/ItemPage.tsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    useCurrentMember();
+
+  if (!itemId) {
+    navigate(HOME_PATH);
+    return null;
+  }
+  return (
+    <ItemLoginWrapper
+      itemId={itemId}
+      item={item}
```
Comment (ReidyT) [inline src/modules/pages/itemPage/ItemPage.tsx:39]: Do we need itemId and item? I haven't check the code, but maybe we could just use item?

File: src/modules/pages/itemPage/RequestAccessContent.tsx [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+} from '@graasp/sdk';
+
+import { Check, Lock } from 'lucide-react';
+
+import { usePlayerTranslation } from '@/config/i18n';
+import { hooks, mutations } from '@/config/queryClient';
+import { REQUEST_MEMBERSHIP_BUTTON_ID } from '@/config/selectors';
+import { PLAYER } from '@/langs/constants';
+
+// eslint-disable-next-line import/prefer-default-export
```
Comment (ReidyT) [inline src/modules/pages/itemPage/RequestAccessContent.tsx:17]: same here?

File: src/modules/pages/itemPage/RequestAccessContent.tsx [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+  if (request?.status === MembershipRequestStatus.Pending) {
+    return (
+      <Stack
+        direction="column"
+        justifyContent="center"
+        alignItems="center"
+        height="100%"
+        gap={2}
+        // data-cy={MEMBERSHIP_REQUEST_PENDING_SCREEN_SELECTOR}
```
Comment (ReidyT) [inline src/modules/pages/itemPage/RequestAccessContent.tsx:41]:
```suggestion
```

File: src/modules/pages/itemPage/RequestAccessContent.tsx [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        {translatePlayer(PLAYER.REQUEST_ACCESS_TITLE)}
+      </Typography>
+      <LoadingButton
+        id={REQUEST_MEMBERSHIP_BUTTON_ID}
+        disabled={isSuccess}
+        loading={isLoading}
+        endIcon={isSuccess ? <Check /> : null}
+        onClick={async () => {
+          await requestMembership({ id: itemId });
+        }}
```
Comment (ReidyT) [inline src/modules/pages/itemPage/RequestAccessContent.tsx:73]: it is necessary to await here?

```suggestion
        onClick={() => requestMembership({ id: itemId })}
```

File: src/modules/pages/itemPage/ItemPage.tsx (hunk 6 above)
Comment (pyphilia) [reply src/modules/pages/itemPage/ItemPage.tsx:35]: A null item might be because you don't have access or you need to item login, it's dealt later in the code.
Here we redirect if the path does not have an id, mostly.

File: src/modules/pages/itemPage/ItemPage.tsx (hunk 7 above)
Comment (pyphilia) [reply src/modules/pages/itemPage/ItemPage.tsx:45]: `ItemLoginWrapper` is in UI. So maybe a solution would be to export selector constants? not sure.

File: src/modules/pages/itemPage/ItemPage.tsx (hunk 8 above)
Comment (pyphilia) [reply src/modules/pages/itemPage/ItemPage.tsx:39]: `item` can be null and it's expected! We need the `id` for every situation, especially when item login.

File: src/modules/pages/itemPage/ItemPage.tsx (hunk 7 above)
Comment (ReidyT) [reply src/modules/pages/itemPage/ItemPage.tsx:45]: I think it's okay like that, I didn't check that came from UI

**Review by ReidyT — APPROVED** [review]
LGTM ☔️

File: cypress/e2e/membershipRequest.cy.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    it('Request membership', () => {
+      cy.intercept(
+        {
+          method: DEFAULT_GET.method,
+          url: new RegExp(
+            `${API_HOST}/items/${ID_FORMAT}/memberships/requests$`,
+          ),
+        },
+        ({ reply }) => {
+          reply([]);
```
Comment (spaenleh) [inline cypress/e2e/membershipRequest.cy.ts:47]: Is this the correct reply ?

File: cypress/e2e/membershipRequest.cy.ts [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+      cy.intercept(
+        {
+          method: DEFAULT_POST.method,
+          url: new RegExp(
+            `${API_HOST}/items/${ID_FORMAT}/memberships/requests$`,
+          ),
+        },
+        ({ reply }) => {
+          reply('ok');
```
Comment (spaenleh) [inline cypress/e2e/membershipRequest.cy.ts:59]: Same here, for the future I think it would be better to return the same thing as the backend. Even if it is only a `StatusCodes.NO_CONTENT`

File: cypress/e2e/pseudonimized.cy.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+      url: new RegExp(`${API_HOST}/items/${ID_FORMAT}/login$`),
+    }).as('itemLoginSignIn');
+
+    const rootItem = FolderItemFactory({ name: 'pseudo' });
+    const items = [rootItem];
+    cy.setUpApi({
+      currentMember: null,
+      items,
+      itemLogins: { [rootItem.id]: `${ItemLoginSchemaType.Username}` },
+    });
```
Comment (spaenleh) [inline cypress/e2e/pseudonimized.cy.ts:28]: Usually we do this at the start of the test, should we move it there to be consistent ?

File: cypress/e2e/pseudonimized.cy.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+      url: new RegExp(`${API_HOST}/items/${ID_FORMAT}/enroll$`),
+    }).as('enroll');
+
+    const rootItem = FolderItemFactory({ name: 'pseudo' });
+    const items = [rootItem];
+    cy.setUpApi({
+      items,
+      itemLogins: { [rootItem.id]: `${ItemLoginSchemaType.Username}` },
+    });
+
```
Comment (spaenleh) [inline cypress/e2e/pseudonimized.cy.ts:55]: Same comment here

File: cypress/fixtures/mockTypes.ts [hunk 16]
```diff
@@ -6,7 +6,7 @@ import {
 } from '@graasp/sdk';
 
 export type MockItemTag = Omit<ItemTag, 'item'>;
-export type MockItem = Omit<PackedItem, 'permission' | 'hidden' | 'public'> & {
```
Comment (spaenleh) [inline cypress/fixtures/mockTypes.ts:9]: Why was this changed ?

File: src/Root.tsx [hunk 17]
```diff
@@ -39,23 +37,39 @@ const globalStyles = (
...
+      <Router>
+        <ErrorBoundary fallback={<FallbackComponent />}>
+          <CurrentMemberContextProvider>
+            <App />
+          </CurrentMemberContextProvider>
+        </ErrorBoundary>
+      </Router>
+    </ThemeProvider>
+  );
+};
```
Comment (spaenleh) [inline src/Root.tsx:63]: I already put a comment about this when it was implemented by Lina, but I do not really like that is is an opaque/black-box component. I would prefer for it to take a children prop and simply wrap it with the required logic.

**Review by spaenleh — COMMENTED** [review]
Looks good to me, I have left a few comments. Let me know.

File: cypress/e2e/pseudonimized.cy.ts (hunk 15 above)
Comment (pyphilia) [reply cypress/e2e/pseudonimized.cy.ts:55]: I think I prefer having the intercept first. In between are the vars necessary to define the setup.

File: cypress/fixtures/mockTypes.ts (hunk 16 above)
Comment (pyphilia) [reply cypress/fixtures/mockTypes.ts:9]: I think we should remove the omit anyway, but I will set back the permission for now.

File: src/Root.tsx (hunk 17 above)
Comment (pyphilia) [reply src/Root.tsx:63]: I revert then, it's because `theme` was deprecated.

File: cypress/e2e/pseudonimized.cy.ts (hunk 15 above)
Comment (spaenleh) [reply cypress/e2e/pseudonimized.cy.ts:55]: okay sounds good

File: cypress/fixtures/mockTypes.ts (hunk 16 above)
Comment (spaenleh) [reply cypress/fixtures/mockTypes.ts]: I was not agains the change, I just could not see why it was done. If it is not necessary to omit permission, then you can keep the change no problem.
